resolver/src/p2p/fetcher.rs

```diff
@@ -264,10 +321,50 @@ impl<E: Clock + GClock + Rng + Metrics, P: PublicKey, Key: Span, NetS: Sender<Pu
     }
 
     /// Blocks a peer from being used to fetch data.
+    ///
+    /// Also removes the peer from all target sets.
     pub fn block(&mut self, peer: P) {
+        // Remove peer from all target sets (keeping empty entries)
+        for targets in self.targets.values_mut() {
+            targets.remove(&peer);
+        }
+
         self.requester.block(peer);
     }
 
+    /// Add or replace target peers for fetching a key.
+    ///
+    /// If `replace` is true, atomically replaces the target set.
+    /// If `replace` is false, adds to the existing target set.
+    ///
+    /// If the resulting target set is empty, the fetch will wait until targets are added.
+    ///
+    /// Clears the waiter to allow immediate retry if the fetch was blocked waiting for targets.
+    pub fn add_targets(&mut self, key: Key, peers: impl IntoIterator<Item = P>, replace: bool) {
```

Comment:
For the initial pass, I'm ok not offering this atomic replace. Once you add targets you can't "undo" (unless a peer is blocked).